src/core/lib/event_engine/grpc_ares_wrapper.cc

```diff
@@ -0,0 +1,1037 @@
+  int status = ares_init_options(&channel_, &opts, ARES_OPT_FLAGS);
+  if (status != ARES_SUCCESS) {
+    gpr_log(GPR_ERROR, "ares_init_options failed, status: %d", status);
+    return AresStatusToAbslStatus(
+        status, absl::StrCat("Failed to init ares channel. c-ares error: ",
+                             ares_strerror(status)));
+  }
+  event_engine_grpc_ares_test_only_inject_config(channel_);
+  // If dns_server is specified, use it.
+  absl::Status error = SetRequestDNSServer(dns_server);
+  if (!error.ok()) {
+    ares_destroy(channel_);
+    return error;
+  }
+  initialized_ = true;
+  return absl::OkStatus();
+}
+
+absl::Status GrpcAresRequest::SetRequestDNSServer(
+    absl::string_view dns_server) {
+  if (!dns_server.empty()) {
+    GRPC_ARES_WRAPPER_TRACE_LOG("request:%p Using DNS server %s", this,
+                                dns_server.data());
+    grpc_resolved_address addr;
+    struct ares_addr_port_node dns_server_addr;
+    if (grpc_parse_ipv4_hostport(dns_server, &addr, /*log_errors=*/false)) {
+      dns_server_addr.family = AF_INET;
+      struct sockaddr_in* in = reinterpret_cast<struct sockaddr_in*>(addr.addr);
+      memcpy(&dns_server_addr.addr.addr4, &in->sin_addr,
+             sizeof(struct in_addr));
+      dns_server_addr.tcp_port = grpc_sockaddr_get_port(&addr);
+      dns_server_addr.udp_port = grpc_sockaddr_get_port(&addr);
+    } else if (grpc_parse_ipv6_hostport(dns_server, &addr,
+                                        /*log_errors=*/false)) {
+      dns_server_addr.family = AF_INET6;
+      struct sockaddr_in6* in6 =
+          reinterpret_cast<struct sockaddr_in6*>(addr.addr);
+      memcpy(&dns_server_addr.addr.addr6, &in6->sin6_addr,
+             sizeof(struct in6_addr));
+      dns_server_addr.tcp_port = grpc_sockaddr_get_port(&addr);
+      dns_server_addr.udp_port = grpc_sockaddr_get_port(&addr);
+    } else {
+      return absl::Status(absl::StatusCode::kInvalidArgument,
+                          absl::StrCat("Cannot parse authority: ", dns_server));
+    }
+    // Prevent an uninitialized variable.
+    dns_server_addr.next = nullptr;
+    int status = ares_set_servers_ports(channel_, &dns_server_addr);
+    if (status != ARES_SUCCESS) {
+      return AresStatusToAbslStatus(
+          status, absl::StrCat("c-ares status is not ARES_SUCCESS: ",
+                               ares_strerror(status)));
+    }
+  }
+  return absl::OkStatus();
+}
+
+void GrpcAresRequest::OnReadable(FdNode* fd_node, absl::Status status) {
+  mu().Lock();
+  GPR_ASSERT(fd_node->readable_registered);
+  fd_node->readable_registered = false;
+  GRPC_ARES_WRAPPER_TRACE_LOG("OnReadable: fd: %d; request: %p; status: %s",
+                              fd_node->as, this, status.ToString().c_str());
+  PrintCurrentStackTrace();
+  if (status.ok() && !shutting_down_) {
+    do {
+      ares_process_fd(channel_, fd_node->as, ARES_SOCKET_BAD);
+    } while (fd_node->polled_fd->IsFdStillReadableLocked());
```

Comment:
Thanks for bringing this up. Is it because our poller (especially epoll1 based poller) is level-triggered so that we need this loop here otherwise we may risk not getting notified again and being stucked?

If having this loop here is essential, then I agree that we should fix this potential issue, maybe something like:
```
fd_set read_fds;
fd_set write_fds;
FD_SET(fd_node->as, &read_fds);
do {
  ares_process_fd(channel_, fd_node->as, ARES_SOCKET_BAD);
  ares_fds(channel_, &read_fds, &write_fds);
} while (FD_ISSET(fd_node->as, &read_fds) && fd_node->polled_fd->IsFdStillReadableLocked());
```

In this case the fd_set semantic seems to be easier to work with than `ares_getsock`. WDYT?

Reply:
Yes, but the loop is needed because otherwise *edge*-triggered pollers won't notice if there are still un-read bytes following `ares_process_fd`.

The `ares_fds` approach looks promising. I think we need to use the `big_fd_set` types that c-ares exposes, for that, though?

Reply:
Oh sorry, ours use edge-triggered. I got mixed-up with the term. Yes, it seems like we need to use `big_fd_set` and include <selectbridge.h>.